pyiron/atomistics/structure/analyse.py

```diff
@@ -0,0 +1,133 @@
+    array([1, 1, 3])
+
+    """
+    labels = np.unique(labels, return_inverse=True)[1]
+    unique_labels = np.unique(labels)
+    mat = coo_matrix((np.ones_like(labels), (labels, np.arange(len(labels)))))
+    mean_values = np.asarray(mat.dot(np.asarray(values).reshape(len(labels), -1))/mat.sum(axis=1))
+    if np.prod(mean_values.shape).astype(int)==len(unique_labels):
+        return mean_values.flatten()
+    return mean_values
+
+class Analyse:
+    """ Class to analyse atom structure.  """
+    def __init__(self, structure):
+        """
+        Args:
+            structure (pyiron.atomistics.structure.atoms.Atoms): reference Atom structure
+        """
+        self._structure = structure
+
+    def get_layers(self, distance_threshold=0.01, id_list=None, wrap_atoms=True, planes=None):
+        """
+        Get an array of layer numbers.
+
+        Args:
+            distance_threshold (float): Distance below which two points are
+                considered to belong to the same layer. For detailed
+                description: sklearn.cluster.AgglomerativeClustering
+            id_list (list/numpy.ndarray): List of atoms for which the layers
+                should be considered.
+            planes (list/numpy.ndarray): Planes along which the layers are calculated. Planes are
+                given in vectors, i.e. [1, 0, 0] gives the layers along the x-axis. Default planes
+                are orthogonal unit vectors: [[1, 0, 0], [0, 1, 0], [0, 0, 1]]. If you have a
+                tilted box and want to calculate the layers along the directions of the cell
+                vectors, use `planes=np.linalg.inv(structure.cell).T`. Whatever values are
+                inserted, they are internally normalized, so whether [1, 0, 0] is entered or
+                [2, 0, 0], the results will be the same.
+
+        Returns: Array of layer numbers (same shape as structure.positions)
+
+        Example I - how to get the number of layers in each direction:
+
+        >>> structure = Project('.').create_structure('Fe', 'bcc', 2.83).repeat(5)
+        >>> print('Numbers of layers:', np.max(structure.analyse.get_layers(), axis=0)+1)
+
+        Example II - get layers of only one species:
+
+        >>> print('Iron layers:', structure.analyse.get_layers(
+        ...       id_list=structure.select_index('Fe')))
+        """
+        if distance_threshold <= 0:
+            raise ValueError('distance_threshold must be a positive float')
+        if id_list is not None and len(id_list)==0:
+            raise ValueError('id_list must contain at least one id')
+        if wrap_atoms and planes is None:
+            positions, indices = self._structure.get_extended_positions(
+                width=distance_threshold, return_indices=True
+            )
+            if id_list is not None:
+                id_list = np.arange(len(self._structure))[np.array(id_list)]
+                id_list = np.any(id_list[:,None]==indices[None,:], axis=0)
```

Comment:
Nit: use np.newaxis

Reply:
It's not the same thing (that was what I thought I discovered recently)

Reply:
But `type(np.newaxis) == NoneType`, so it must be the same, no?

Reply:
Yes so I thought it doesn’t matter which one I use.

Reply:
Yes, it doesn't matter for the code, but `np.newaxis` is documenting your intent.

Reply:
ok :D